File: src/rust/iced-x86-py/tests/decoder_test.py
# SPDX-License-Identifier: MIT
# Copyright (C) 2018-present iced project and contributors

import copy
import pytest
from iced_x86 import *

@pytest.mark.parametrize("bitness", [16, 32, 64, 0, 15, 128])
def test_invalid_bitness(bitness):
	data = b"\x90"
	if bitness == 16 or bitness == 32 or bitness == 64:
		Decoder(bitness, data)
	else:
		with pytest.raises(ValueError):
			Decoder(bitness, data)

@pytest.mark.parametrize("data, is_valid", [
	(b"", True),
	(b"\x90", True),
	(bytearray(b""), True),
	(bytearray(b"\x90"), True),
	(memoryview(b""), False),
	(memoryview(b"\x90"), False),
	(memoryview(bytearray(b"")), False),
	(memoryview(bytearray(b"\x90")), False),
	(None, False),
	(123, False),
	("Hello", False),
])
def test_different_data_types(data, is_valid):
	bitness = 64
	if is_valid:
		decoder = Decoder(bitness, data)
		if len(data) == 0:
			assert decoder.decode().code == Code.INVALID
			assert decoder.last_error == DecoderError.NO_MORE_BYTES
		else:
			assert decoder.decode().code != Code.INVALID
			assert decoder.last_error == DecoderError.NONE
	else:
		with pytest.raises(TypeError):
			Decoder(bitness, data)

@pytest.mark.parametrize("bitness, data, code", [
	(16, b"\x03\xCE", Code.ADD_R16_RM16),
	(32, b"\x03\xCE", Code.ADD_R32_RM32),
	(64, b"\x48\x03\xCE", Code.ADD_R64_RM64),
])
def test_bitness(bitness, data, code):
	decoder = Decoder(bitness, data)
	assert decoder.bitness == bitness
	instr = decoder.decode()
	assert instr.code == code

def test_without_decoder_options():
	decoder = Decoder(64, b"\x0F\x1A\x08")
	instr = decoder.decode()
	assert instr.code != Code.BNDLDX_BND_MIB

def test_with_decoder_options():
	decoder = Decoder(64, b"\x0F\x1A\x08", DecoderOptions.MPX)
	instr = decoder.decode()
	assert instr.code == Code.BNDLDX_BND_MIB

def test_iter():
	decoder = Decoder(64, b"\x48\x09\xCE\x90\xF3\x90")
	instrs = [instr for instr in decoder]
	assert len(instrs) == 3
	assert instrs[0].code == Code.OR_RM64_R64
	assert instrs[1].code == Code.NOPD
	assert instrs[2].code == Code.PAUSE

def test_ip():
	decoder = Decoder(64, b"\x48\x09\xCE\x90")
	assert decoder.ip == 0
	decoder = Decoder(64, b"\x48\x09\xCE\x90", ip=0xABCD_EF01_2345_6789)
	assert decoder.ip == 0xABCD_EF01_2345_6789
	decoder.ip = 0x1234_5678_9ABC_DEF0
	assert decoder.ip == 0x1234_5678_9ABC_DEF0
	decoder.decode()
	assert decoder.ip == 0x1234_5678_9ABC_DEF3
	decoder.decode()
	assert decoder.ip == 0x1234_5678_9ABC_DEF4
	decoder.decode()
	assert decoder.ip == 0x1234_5678_9ABC_DEF4
	decoder.ip = 0xFEDC_BA98_7654_3210
	assert decoder.ip == 0xFEDC_BA98_7654_3210

def test_position():
	decoder = Decoder(64, b"\x48\x09\xCE\x90\xF3\x90")

	for _ in range(2):
		assert decoder.can_decode
		assert decoder.max_position == 6
		assert decoder.position == 0

		assert decoder.decode().code == Code.OR_RM64_R64
		assert decoder.can_decode
		assert decoder.max_position == 6
		assert decoder.position == 3

		assert decoder.decode().code == Code.NOPD
		assert decoder.can_decode
		assert decoder.max_position == 6
		assert decoder.position == 4

		assert decoder.decode().code == Code.PAUSE
		assert not decoder.can_decode
		assert decoder.max_position == 6
		assert decoder.position == 6

		assert decoder.decode().code == Code.INVALID
		assert not decoder.can_decode
		assert decoder.max_position == 6
		assert decoder.position == 6

		decoder.position = 0

	decoder.position = 3
	assert decoder.decode().code == Code.NOPD
	assert decoder.can_decode
	assert decoder.max_position == 6
	assert decoder.position == 4

def test_invalid_position():
	decoder = Decoder(64, b"\x48\x09\xCE\x90\xF3\x90")
	decoder.position = 0
	decoder.position = 6
	with pytest.raises(ValueError):
		decoder.position = 7

def test_last_error():
	decoder = Decoder(64, b"\x90\x48")
	assert decoder.last_error == DecoderError.NONE

	assert decoder.decode().code == Code.NOPD
	assert decoder.last_error == DecoderError.NONE
	assert decoder.decode().code == Code.INVALID
	assert decoder.last_error == DecoderError.NO_MORE_BYTES

	decoder = Decoder(64, b"\xF0\x90")
	assert decoder.decode().code == Code.INVALID
	assert decoder.last_error == DecoderError.INVALID_INSTRUCTION

def test_decode():
	data = b"\x48\x09\xCE\x90\xF3\x90"
	decodera = Decoder(64, data)
	decoderb = Decoder(64, data)
	instr = Instruction()

	instr1 = decodera.decode()
	decoderb.decode_out(instr)
	assert instr.eq_all_bits(instr1)

	instr2 = decodera.decode()
	decoderb.decode_out(instr)
	assert instr.eq_all_bits(instr2)

def test_offsets():
	decoder = Decoder(64, b"\x90\x83\xB3\x34\x12\x5A\xA5\x5A")

	instr = decoder.decode()
	co = decoder.get_constant_offsets(instr)
	assert co.displacement_offset == 0
	assert co.displacement_size == 0
	assert co.immediate_offset == 0
	assert co.immediate_size == 0
	assert co.immediate_offset2 == 0
	assert co.immediate_size2 == 0
	assert not co.has_displacement
	assert not co.has_immediate
	assert not co.has_immediate2

	instr = decoder.decode()
	co = decoder.get_constant_offsets(instr)
	assert co.displacement_offset == 2
	assert co.displacement_size == 4
	assert co.immediate_offset == 6
	assert co.immediate_size == 1
	assert co.immediate_offset2 == 0
	assert co.immediate_size2 == 0
	assert co.has_displacement
	assert co.has_immediate
	assert not co.has_immediate2

def test_co_eq_ne_hash():
	decoder = Decoder(64, b"\x90\x90\x83\xB3\x34\x12\x5A\xA5\x5A\x83\xB3\x34\x12\x5A\xA5\x5A")
	instr = decoder.decode()
	co1 = decoder.get_constant_offsets(instr)
	instr = decoder.decode()
	co2 = decoder.get_constant_offsets(instr)
	instr = decoder.decode()
	co3 = decoder.get_constant_offsets(instr)
	instr = decoder.decode()
	co4 = decoder.get_constant_offsets(instr)

	assert id(co1) != id(co2)
	assert id(co3) != id(co4)

	assert hash(co1) == hash(co2)
	assert hash(co3) == hash(co4)

	assert co1 == co2
	assert not (co1 != co2)

	assert co3 == co4
	assert not (co3 != co4)

	assert co1 != co3
	assert not (co1 == co3)

	assert co1 != 1
	assert co1 != 1.23
	assert co1 != None
	assert co1 != []
	assert co1 != {}
	assert co1 != (1, 2)

	assert not (co1 == 1)
	assert not (co1 == 1.23)
	assert not (co1 == None)
	assert not (co1 == [])
	assert not (co1 == {})
	assert not (co1 == (1, 2))

@pytest.mark.parametrize("copy_co", [
	lambda instr: copy.copy(instr),
	lambda instr: copy.deepcopy(instr),
	lambda instr: instr.copy(),
])
def test_co_copy_deepcopy_mcopy(copy_co):
	decoder = Decoder(64, b"\x90\x83\xB3\x34\x12\x5A\xA5\x5A", ip=0x1234_5678_9ABC_DEF1)
	coa = decoder.get_constant_offsets(decoder.decode())
	cob = decoder.get_constant_offsets(decoder.decode())

	coa2 = copy_co(coa)
	assert coa is not coa2
	assert id(coa) != id(coa2)
	assert coa == coa2

	cob2 = copy_co(cob)
	assert cob is not cob2
	assert id(cob) != id(cob2)
	assert cob == cob2

def test_no_bytes():
	decoder = Decoder(64, b"")
	assert not decoder.can_decode
	assert decoder.position == 0
	assert decoder.max_position == 0

	instr = decoder.decode()
	assert decoder.last_error == DecoderError.NO_MORE_BYTES
	assert instr.is_invalid
	assert not instr
	assert not decoder.can_decode
	assert decoder.position == 0
	assert decoder.max_position == 0

def test_no_invalid_check_option():
	data = b"\xF0\x02\xCE"
	decodera = Decoder(64, data)
	decoderb = Decoder(64, data, DecoderOptions.NO_INVALID_CHECK)

	instra = decodera.decode()
	instrb = decoderb.decode()

	assert instra.code == Code.INVALID
	assert decodera.last_error == DecoderError.INVALID_INSTRUCTION
	assert instrb.code == Code.ADD_R8_RM8

def test_amd_option():
	data = b"\x66\x70\x5A"
	decodera = Decoder(64, data)
	decoderb = Decoder(64, data, DecoderOptions.AMD)

	instra = decodera.decode()
	instrb = decoderb.decode()

	assert instra.code == Code.JO_REL8_64
	assert instrb.code == Code.JO_REL8_16

def test_multiple_options():
	decoder = Decoder(64, b"\xF3\x90", DecoderOptions.AMD | DecoderOptions.NO_PAUSE)
	instr = decoder.decode()
	assert instr.code == Code.NOPD

File: src/rust/iced-x86-js/src/tests/decoder.test.js
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

const { Code, Decoder, DecoderError, DecoderOptions, Instruction, OpKind, Register } = require("iced-x86");

test("Create a Decoder with no bytes", () => {
	const bytes = new Uint8Array([]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);
	expect(decoder.canDecode).toBe(false);
	const instr = decoder.decode();
	expect(instr.isInvalid).toBe(true);
	expect(instr.code).toBe(Code.INVALID);
	expect(decoder.lastError).toBe(DecoderError.NoMoreBytes);
	instr.free();
	decoder.free();
});

test("Create a 16-bit Decoder", () => {
	const bytes = new Uint8Array([0x01, 0x18]);
	const decoder = new Decoder(16, bytes, DecoderOptions.None);

	expect(decoder.bitness).toBe(16);
	const instr = decoder.decode();
	expect(instr.isInvalid).toBe(false);
	expect(instr.code).toBe(Code.Add_rm16_r16);
	expect(instr.op0Kind).toBe(OpKind.Memory);
	expect(instr.memoryBase).toBe(Register.BX);
	expect(instr.memoryIndex).toBe(Register.SI);

	decoder.free();
	instr.free();
});

test("Create a 32-bit Decoder", () => {
	const bytes = new Uint8Array([0x01, 0x18]);
	const decoder = new Decoder(32, bytes, DecoderOptions.None);

	expect(decoder.bitness).toBe(32);
	const instr = decoder.decode();
	expect(instr.isInvalid).toBe(false);
	expect(instr.code).toBe(Code.Add_rm32_r32);
	expect(instr.op0Kind).toBe(OpKind.Memory);
	expect(instr.memoryBase).toBe(Register.EAX);
	expect(instr.memoryIndex).toBe(Register.None);

	decoder.free();
	instr.free();
});

test("Create a 64-bit Decoder", () => {
	const bytes = new Uint8Array([0x48, 0x01, 0x18]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);

	expect(decoder.bitness).toBe(64);
	const instr = decoder.decode();
	expect(instr.isInvalid).toBe(false);
	expect(instr.code).toBe(Code.Add_rm64_r64);
	expect(instr.op0Kind).toBe(OpKind.Memory);
	expect(instr.memoryBase).toBe(Register.RAX);
	expect(instr.memoryIndex).toBe(Register.None);

	decoder.free();
	instr.free();
});

test("Decode multiple instructions", () => {
	const bytes = new Uint8Array([0x86, 0x64, 0x32, 0x16, 0xF0, 0xF2, 0x83, 0x00, 0x5A]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);

	expect(decoder.maxPosition).toBe(9);
	expect(decoder.canDecode).toBe(true);
	expect(decoder.position).toBe(0);
	const instr1 = decoder.decode();
	expect(instr1.code).toBe(Code.Xchg_rm8_r8);
	expect(decoder.canDecode).toBe(true);
	expect(decoder.position).toBe(4);
	const instr2 = decoder.decode();
	expect(instr2.code).toBe(Code.Add_rm32_imm8);
	expect(decoder.canDecode).toBe(false);
	expect(decoder.position).toBe(9);
	const instr3 = decoder.decode();
	expect(instr3.code).toBe(Code.INVALID);
	expect(decoder.canDecode).toBe(false);
	expect(decoder.position).toBe(9);

	decoder.free();
	instr1.free();
	instr2.free();
	instr3.free();
});

test("Decode with decodeOut()", () => {
	const bytes = new Uint8Array([0x86, 0x64, 0x32, 0x16, 0xF0, 0xF2, 0x83, 0x00, 0x5A, 0x48, 0x89, 0xCE]);
	const decodera = new Decoder(64, bytes, DecoderOptions.None);
	const decoderb = new Decoder(64, bytes, DecoderOptions.None);
	const instr = new Instruction();

	const instr1 = decodera.decode();
	decoderb.decodeOut(instr);
	expect(instr1.equals(instr)).toBe(true);

	const instr2 = decodera.decode();
	decoderb.decodeOut(instr);
	expect(instr2.equals(instr)).toBe(true);

	const instr3 = decodera.decode();
	decoderb.decodeOut(instr);
	expect(instr3.equals(instr)).toBe(true);

	const instr4 = decodera.decode();
	decoderb.decodeOut(instr);
	expect(instr4.equals(instr)).toBe(true);

	decodera.free();
	decoderb.free();
	instr.free();
	instr1.free();
	instr2.free();
	instr3.free();
	instr4.free();
});

test("Decode all instructions", () => {
	const bytes = new Uint8Array([0x86, 0x64, 0x32, 0x16, 0xF0, 0xF2, 0x83, 0x00, 0x5A, 0x48, 0x89, 0xCE]);
	const decodera = new Decoder(64, bytes, DecoderOptions.None);
	const decoderb = new Decoder(64, bytes, DecoderOptions.None);
	const decoderc = new Decoder(64, bytes, DecoderOptions.None);

	const instrs = decoderb.decodeInstructions(0);
	expect(instrs.length).toBe(0);
	const instrsa = decodera.decodeAll();
	const instrsb = decoderb.decodeInstructions(0xFFFFFFFF);
	expect(decodera.canDecode).toBe(false);
	expect(decoderb.canDecode).toBe(false);
	expect(instrsa.length).toBe(3);
	expect(instrsb.length).toBe(3);
	for (let i = 0; i < instrsa.length; i++)
		expect(instrsa[i].equals(instrsb[i])).toBe(true);
	const instrsc = decodera.decodeAll();
	const instrsd = decoderb.decodeInstructions(0xFFFFFFFF);
	expect(instrsc.length).toBe(0);
	expect(instrsd.length).toBe(0);

	const instrse = decoderc.decodeInstructions(2);
	const instrsf = decoderc.decodeInstructions(3);
	expect(instrse.length).toBe(2);
	expect(instrsf.length).toBe(1);
	for (let i = 0; i < 2; i++)
		expect(instrsa[i].equals(instrse[i])).toBe(true);
	for (let i = 2; i < 3; i++)
		expect(instrsa[i].equals(instrsf[i - 2])).toBe(true);

	decodera.free();
	decoderb.free();
	decoderc.free();
	instrs.forEach(a => a.free());
	instrsa.forEach(a => a.free());
	instrsb.forEach(a => a.free());
	instrsc.forEach(a => a.free());
	instrsd.forEach(a => a.free());
	instrse.forEach(a => a.free());
	instrsf.forEach(a => a.free());
});

test("Decode with DecoderOptions.NoInvalidCheck", () => {
	const bytes = new Uint8Array([0xF0, 0x02, 0xCE]);
	const decodera = new Decoder(64, bytes, DecoderOptions.None);
	const decoderb = new Decoder(64, bytes, DecoderOptions.NoInvalidCheck);

	const instra = decodera.decode();
	const instrb = decoderb.decode();
	expect(instra.code).toBe(Code.INVALID);
	expect(decodera.lastError).toBe(DecoderError.InvalidInstruction);
	expect(instrb.code).toBe(Code.Add_r8_rm8);

	decodera.free();
	decoderb.free();
	instra.free();
	instrb.free();
});

test("Decode with DecoderOptions.AMD", () => {
	const bytes = new Uint8Array([0x66, 0x70, 0x5A]);
	const decodera = new Decoder(64, bytes, DecoderOptions.None);
	const decoderb = new Decoder(64, bytes, DecoderOptions.AMD);

	const instra = decodera.decode();
	const instrb = decoderb.decode();
	expect(instra.code).toBe(Code.Jo_rel8_64);
	expect(instrb.code).toBe(Code.Jo_rel8_16);

	decodera.free();
	decoderb.free();
	instra.free();
	instrb.free();
});

test("Creating a Decoder with an invalid bitness throws", () => {
	expect(() => new Decoder(63, new Uint8Array([0x90]), DecoderOptions.None)).toThrow();
});

// Make sure it's not an enum arg in the Rust code since it's a flags enum. It must be a u32 in the method sig.
test("Create a Decoder with multiple options", () => {
	const decoder = new Decoder(64, new Uint8Array([0xF3, 0x90]), DecoderOptions.AMD | DecoderOptions.NoPause);
	const instr = decoder.decode();
	expect(instr.code).toBe(Code.Nopd);
	decoder.free();
	instr.free();
});

test("Decoder.IP prop", () => {
	const decoder = new Decoder(64, new Uint8Array([0x90]), DecoderOptions.None);
	expect(decoder.ip).toBe(0n);
	decoder.ip = 0x12345678n;
	expect(decoder.ip).toBe(0x12345678n);
	decoder.ip = 0x9ABCDEFD12345678n;
	expect(decoder.ip).toBe(0x9ABCDEFD12345678n);
	decoder.free();
});

test("Set Decoder position", () => {
	const bytes = new Uint8Array([0x86, 0x64, 0x32, 0x16, 0xF0, 0xF2, 0x83, 0x00, 0x5A, 0x48, 0x89, 0xCE]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);

	const instrs1 = decoder.decodeAll();
	const instrs2 = decoder.decodeAll();
	decoder.position = 0;
	const instrs3 = decoder.decodeAll();
	const instrs4 = decoder.decodeAll();
	decoder.position = 4;
	const instrs5 = decoder.decodeAll();
	const instrs6 = decoder.decodeAll();

	expect(instrs2.length).toBe(0);
	expect(instrs4.length).toBe(0);
	expect(instrs6.length).toBe(0);
	expect(instrs1.length).toBe(3);
	expect(instrs3.length).toBe(3);
	expect(instrs5.length).toBe(2);
	for (let i = 0; i < instrs1.length; i++) {
		expect(instrs1[i].equals(instrs3[i]));
		if (i >= 1)
			expect(instrs1[i].equals(instrs5[i - 1]));
	}

	decoder.free();
	instrs1.forEach(a => a.free());
	instrs2.forEach(a => a.free());
	instrs3.forEach(a => a.free());
	instrs4.forEach(a => a.free());
	instrs5.forEach(a => a.free());
	instrs6.forEach(a => a.free());
});

test("Decoder lastError prop", () => {
	const bytes = new Uint8Array([0xF0, 0x00, 0xCE, 0xF3]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);

	const instr1 = decoder.decode();
	expect(instr1.code).toBe(Code.INVALID);
	expect(decoder.lastError).toBe(DecoderError.InvalidInstruction);
	const instr2 = decoder.decode();
	expect(instr2.code).toBe(Code.INVALID);
	expect(decoder.lastError).toBe(DecoderError.NoMoreBytes);

	decoder.free();
	instr1.free();
	instr2.free();
});

test("Decoder constant offsets", () => {
	const bytes = new Uint8Array([0x90, 0x83, 0xB3, 0x34, 0x12, 0x5A, 0xA5, 0x5A]);
	const decoder = new Decoder(64, bytes, DecoderOptions.None);

	const instr1 = decoder.decode();
	const co1 = decoder.getConstantOffsets(instr1);
	expect(co1.hasDisplacement).toBe(false);
	expect(co1.displacementOffset).toBe(0);
	expect(co1.displacementSize).toBe(0);
	expect(co1.hasImmediate).toBe(false);
	expect(co1.immediateOffset).toBe(0);
	expect(co1.immediateSize).toBe(0);
	expect(co1.hasImmediate2).toBe(false);
	expect(co1.immediateOffset2).toBe(0);
	expect(co1.immediateSize2).toBe(0);

	const instr2 = decoder.decode();
	const co2 = decoder.getConstantOffsets(instr2);
	expect(co2.hasDisplacement).toBe(true);
	expect(co2.displacementOffset).toBe(2);
	expect(co2.displacementSize).toBe(4);
	expect(co2.hasImmediate).toBe(true);
	expect(co2.immediateOffset).toBe(6);
	expect(co2.immediateSize).toBe(1);
	expect(co2.hasImmediate2).toBe(false);
	expect(co2.immediateOffset2).toBe(0);
	expect(co2.immediateSize2).toBe(0);

	decoder.free();
	instr1.free();
	instr2.free();
	co1.free();
	co2.free();
});
